scripts/add_electricity.py

```diff
@@ -386,6 +394,93 @@ def attach_storage(
     return
 
 
+def add_retirements(n):
+    if isinstance(retirements_df, pd.DataFrame):
+        
+        carriers = retirements_df.columns
+        
+        df = retirements_df.loc[model_years, :].cumsum()
+        c_by_carrier = n.generators.groupby('carrier').sum().loc[carriers, 'p_nom']
+        
+        for carrier in carriers:
+            existing_cap = c_by_carrier.loc[carrier]
+            for year in model_years:
+                try:
+                    retirement = df.loc[year, carrier]
+                    remaining = max(existing_cap - retirement, 0)
+                    
+                    limit = remaining * (8760/resolution)
+                    
+                    n.add(class_name="GlobalConstraint",
+                            name=f'{carrier} Energy Limit {year}',
+                            sense='<=',
+                            carrier_attribute=carrier,
+                            constant=limit,
+                            type='operational_limit',
+                            investment_period=year,
+                            )
+                except (AttributeError, KeyError, TypeError):
+                    pass
+                
+    return
+
+
+def add_capacity_max(n):
+    if isinstance(capacity_limits_df, pd.DataFrame):
+        
+        carriers = capacity_limits_df.columns
+        
+        df = capacity_limits_df.copy()
+        n_buses = n.buses.shape[0]
+        
+        for carrier in carriers:
+            for year in df.index:
+                try:
+                    limit = df.loc[year, carrier] / n_buses
+                    n.add(class_name="GlobalConstraint",
+                            name=f'{carrier} Capacity Limit {year}',
+                            sense='<=',
+                            carrier_attribute=carrier,
+                            constant=limit,
+                            type='tech_capacity_expansion_limit',
+                            investment_period=year,
+                            )
+                except (AttributeError, KeyError, TypeError):
+                    pass
+            
+    return
+
+def add_energy_max(n):
+    """
+    This function applies an artificial "capacity max" by limiting
+    the amount of energy that a particular carrier can produce in 
+    a given year.
+    """
+    if isinstance(capacity_limits_df, pd.DataFrame):
+        
+        carriers = capacity_limits_df.columns
+        
+        df = capacity_limits_df.copy()
+        
+        for carrier in carriers:
+            for year in df.index:
+                try:
+                    capacity_lim = df.loc[year, carrier]
+                    energy_lim = capacity_lim*(8760/resolution)
+                    name = f'{carrier} Energy Limit 2025'
```

Comment:
Shouldn't year be a variable year, rather than fixed at 2025? Per the docstring this function applies limits for a given year, not a fixed year.